### ISSUE TITLE
Potential serious bug in libs/nets/pyramid_network.py?

### ISSUE THREAD

**Issue by xxxrrr** [CharlesShang/FastMaskRCNN#68]
Line 252-265: from rois to cropped_rois, the order of rois has been changed, which means the outputs['refined’][‘box’] has different order compared to rois. However, bbox_targets (Line 420) has the same order as rois. So for Line 441 refined_box_loss = bbox_inside_weights * _smooth_l1_dist(boxes, bbox_targets), there is misalignment between boxes and xbox_targets. If this true, this is a serious bug. Or am I misunderstanding something here? Thanks.

**Comment by souryuu** [comment]
I agree with you. The order is changed and could lead to serious performance drop. 

@amirbar: You probably wanna look into this.

**Comment by crazylyf** [comment]
Is the problem solved?

**Comment by souryuu** [comment]
No. I have tried to change the order of rois to match with cropped_rois but it still cannot train the refined boxes and classes. The values from refined boxes still close to zero and the network still cannot predict anything else but class 0 and 1(There should be 81 classes predicted).

**Comment by realwecan** [comment]
@souryuu Do you know why the network cannot predict anything but classes 0 and 1? Thanks!

**Comment by souryuu** [comment]
I made several changes and got better classification results. 
1. Set resnet to be trainable and correct the size of image. @amirbar has already fixed this in his recent commit #50.
2. As mentioned above, match the order of rois and crop_rois. 
3. Add batch normalization. I changed all the _extra_conv_conv_scope in pyramid_network.py to _extra_conv_arg_scope_with_bn.

Changes in 1 and 2 seem to be critical. I did not compare the version with and without the changes in 3 yet but it made the trained network to produce decent (more colorful?) cropped_rois.

**Comment by amirbar** [comment]
@souryuu can you please come up with a PR?

On Jun 24, 2017 5:22 AM, "souryuu" <notifications@github.com> wrote:

I made several changes and got better classification results.

   1. Set resnet to be trainable and correct the size of image. @amirbar
   <https://github.com/amirbar> has already fixed this in his recent commit
   #50 <https://github.com/CharlesShang/FastMaskRCNN/pull/50>.
   2. As mentioned above, match the order of rois and crop_rois.
   3. Add batch normalization. I changed all the _extra_conv_conv_scope in
   pyramid_network.py to _extra_conv_arg_scope_with_bn.

Changes in 1 and 2 seem to be critical. I did not compare the version with
and without the changes in 3 yet but it made the trained network to produce
decent (more colorful?) cropped_rois.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/CharlesShang/FastMaskRCNN/issues/68#issuecomment-310807563>,
or mute the thread

**Comment by souryuu** [comment]
@amirbar #96 
Since I changed many parts of the code for visualization, it is quite messy.

**Comment by HuangBo-Terraloupe** [comment]
@souryuu 
Is your code already merged in the master branch? 
Because I trained the network for 2 days, the results seems really bad. 
And I still do not know how to write a evaluation code. (how to you visualize the output)
can you please help and share your code?

thanks
